hpx/runtime/actions/continuation.hpp

```diff
@@ -925,6 +939,14 @@ namespace hpx { namespace actions
     template <typename Arg0>
     void continuation::trigger(Arg0 && arg0)
     {
+        // The dynamic cast decays the argument type to avoid the assert firing
+        // for cases when Arg0 is a const&. This does not make the code invalid
+        // as trigger_value (which is a virtual function) takes its argument
+        // by && anyways.
+        HPX_ASSERT(0 != dynamic_cast<
+                typed_continuation<typename util::decay<Arg0>::type> *
+            >(this));
+
```

Comment:
> Shouldn't it be `typed_continuation<id_type>` in the first place?

That's what I said. `this` points to an object of that type.

> Do we really need forwarding references here? Aren't rvalue references enough? `set_lco_value` accepts rvalue references iirc.

That does not work if the function is called with a `const&`, didn't I mention that?

The easiest would be if you changed the static_cast to use the decayed type yourself. That would demonstrate the problem right away.